### PR TITLE
Refactoring amcedns api

### PR DESCRIPTION
**Pull request by dampfklon**
Account data moved from account to domain config as this is the indented mode of operation for acmedns [ref ](https://github.com/joohoi/acme-dns/issues/76#issuecomment-385149688)

acmedns accounts are created on creation so no unnecessary manual work, but support for the old mode of operation.

### CODE REVIEW

**Comment by dampfklon** [comment]
Do I have to change something?

File: dnsapi/dns_acmedns.sh [hunk 1]
```diff
@@ -1,31 +1,62 @@
...
+#Report Bugs here: https://github.com/dampfklon/acme.sh
+#
+# Usage:
+# export ACMEDNS_BASE_URL="https://auth.acme-dns.io"
+#
+# You can optionally define an already existing account:
+#
+# export ACMEDNS_USERNAME="https://auth.acme-dns.io"
+# export ACMEDNS_PASSWORD="https://auth.acme-dns.io"
+# export ACMEDNS_SUBDOMAIN="https://auth.acme-dns.io"
```
Comment (wilsonianb) [inline dnsapi/dns_acmedns.sh]:
```
# export ACMEDNS_USERNAME="<username>"
# export ACMEDNS_PASSWORD="<password>"
# export ACMEDNS_SUBDOMAIN="<subdomain>"
```

**Comment by Daniel15** [comment]
@Neilpang - Are you going to merge this?

File: dnsapi/dns_acmedns.sh (hunk 1 above)
Comment (Neilpang) [reply dnsapi/dns_acmedns.sh]: can you fix these typos ?

File: dnsapi/dns_acmedns.sh [hunk 2]
```diff
@@ -1,31 +1,62 @@
...
-  ACMEDNS_SUBDOMAIN="${ACMEDNS_SUBDOMAIN:-$(_readaccountconf_mutable ACMEDNS_SUBDOMAIN)}"
+  ACMEDNS_UPDATE_URL="$ACMEDNS_BASE_URL/update"
+  ACMEDNS_REGISTER_URL="$ACMEDNS_BASE_URL/register"
 
-  if [ "$ACMEDNS_UPDATE_URL" = "" ]; then
-    ACMEDNS_UPDATE_URL="https://auth.acme-dns.io/update"
+  if [ -z "$ACMEDNS_USERNAME" ] || [ -z "$ACMEDNS_PASSWORD" ]; then
+    response="$(_post "" "$ACMEDNS_REGISTER_URL" "" "POST")"
+    _debug response "$response"
+    ACMEDNS_USERNAME=$(echo "$response" | sed -E 's/^\{.*?\"username\":\"([^\"]*)\".*\}/\1/g;t;d')
```
Comment (Neilpang) [inline dnsapi/dns_acmedns.sh]: do not use `sed -E`,  please use `_egrep_o(), sed or cut` instead.

File: dnsapi/dns_acmedns.sh [hunk 3]
```diff
@@ -1,31 +1,62 @@
...
+    _info "##########################################################"
+    _info "Press any key to continue... "
+    read -r
   fi
 
-  _saveaccountconf_mutable ACMEDNS_UPDATE_URL "$ACMEDNS_UPDATE_URL"
-  _saveaccountconf_mutable ACMEDNS_USERNAME "$ACMEDNS_USERNAME"
-  _saveaccountconf_mutable ACMEDNS_PASSWORD "$ACMEDNS_PASSWORD"
-  _saveaccountconf_mutable ACMEDNS_SUBDOMAIN "$ACMEDNS_SUBDOMAIN"
+  _saveaccountconf_mutable ACMEDNS_BASE_URL "$ACMEDNS_BASE_URL"
```
Comment (Neilpang) [inline dnsapi/dns_acmedns.sh]: do not save such  ACMEDNS_USERNAME variables,  unless it's exported by user at commandline.

Comment (dampfklon) [reply dnsapi/dns_acmedns.sh]: It is required to save ACMEDNS_USERNAME and ACMEDNS_PASSWORD those are resemble a API Token for the ACME-DNS service. On default operaration those are requested from acmedns-api on first run. You only need to export them if you want to use an existing account which is discouraged as you should use one account per domain.

**Comment by dampfklon** [comment]
@Neilpang are you ok with the changes?

**Comment by Wimmesberger** [comment]
i tested this PR on my environment (ubuntu 18.04) and found that due acme.sh running with the sh shell the `read -r` results in this error: 
`/usr/bin/acme.sh: 53: read: arg count`

read in sh requires a variable to save the input to like
`read -r some_variable_name_that_does_not_occur_anywhere_else`

**Comment by dampfklon** [comment]
@Wimmesberger thanks for the report, fixed

**Comment by Carbenium** [comment]
@Neilpang  Could we get this merged? Without ```_*domainconf``` usage this module just half as useful as it could be.

**Comment by Neilpang** [comment]
resolve the conflicts first.

**Comment by dampfklon** [comment]
Conflicts resolved. Will adapt wiki when merged

**Comment by Carbenium** [comment]
@Neilpang just a shy ping

**Comment by 0xFelix** [comment]
> @Neilpang just a shy ping

Another shy ping

File: dnsapi/dns_acmedns.sh [hunk 4]
```diff
@@ -1,31 +1,62 @@
...
 dns_acmedns_add() {
   fulldomain=$1
   txtvalue=$2
   _info "Using acme-dns"
-  _debug fulldomain "$fulldomain"
-  _debug txtvalue "$txtvalue"
+  _debug "fulldomain $fulldomain"
+  _debug "txtvalue $txtvalue"
+
+  ACMEDNS_BASE_URL="${ACMEDNS_BASE_URL:-$(_readaccountconf_mutable ACMEDNS_BASE_URL)}"
```
Comment (Neilpang) [inline dnsapi/dns_acmedns.sh:27]: Why is ACMEDNS_BASE_URL still saved in accountconf ?

Comment (dampfklon) [reply dnsapi/dns_acmedns.sh:27]: The idea is all acmesh Domains using acme-dns use the same acme-dns server. But each domain has its own account which it uses to publish the challenge.

Comment (Neilpang) [reply dnsapi/dns_acmedns.sh:27]: it looks not good to me.
I prefer that we save all the values in the same place.  no matter in domainconf or in accountconf.

Comment (Neilpang) [reply dnsapi/dns_acmedns.sh:27]: any update?

Comment (dampfklon) [reply dnsapi/dns_acmedns.sh:27]: From the intendet mode of operation the seperation makes sense for me. If you dont like the idea, I can duplicate the ACMEDNS_BASE_URL to all domain configurations.

Comment (Neilpang) [reply dnsapi/dns_acmedns.sh:27]: that will be great.
Thanks

**Comment by 0xFelix** [comment]
@Neilpang I think this can be closed since #3231 was merged.

**Pull request closed without merging**